include/thread_pool/work_stealing_deque.h

```diff
@@ -0,0 +1,205 @@
+#pragma once
+
+#include <atomic>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace dp {
+
+#ifdef __cpp_lib_hardware_interference_size
+    using std::hardware_destructive_interference_size;
+#else
+    // 64 bytes on x86-64 │ L1_CACHE_BYTES │ L1_CACHE_SHIFT │ __cacheline_aligned │ ...
+    inline constexpr std::size_t hardware_destructive_interference_size =
+        2 * sizeof(std::max_align_t);
+#endif
+
+    /**
+     * @brief Chase-Lev work stealing queue
+     * @details Support single producer, multiple consumer. The producer owns the back, consumers
+     * own the top. Consumers can also take from the top of the queue. The queue is "lock-free" in
+     * that it does not directly use mutexes or locks.
+     *
+     * This is an implementation of the deque described in "Correct and Efficient Work-Stealing for
+     * Weak Memory Models" and "Dynamic Circular Work-Stealing Deque" by Chase,Lev.
+     *
+     * This implementation is taken from the following implementations
+     * - https://github.com/ConorWilliams/ConcurrentDeque/blob/main/include/riften/deque.hpp
+     * - https://github.com/taskflow/work-stealing-queue/blob/master/wsq.hpp
+     *
+     * I've made some minor edits and changes based on new C++ 23 features and other personal coding
+     * style choices/preferences.
+     */
+    template <typename T>
+        requires std::is_destructible_v<T>
+    class work_stealing_deque final {
+        /**
+         * @brief Simple circular array buffer that can regrow
+         */
+        class circular_buffer final {
+            std::int64_t size_;
+            std::int64_t mask_;
+            std::unique_ptr<T[]> buffer_ = std::make_unique_for_overwrite<T[]>(size_);
+
+          public:
+            explicit circular_buffer(const std::int64_t size) : size_(size), mask_(size - 1) {
+                // size must be a power of 2
+                assert((size % 2) == 0);
```

Comment:
The current code doesn't actually check for a power of two, contrary to the comment (ex: if size == 6, the assert will pass although that isn't a power of two)
```suggestion
                assert(std::has_single_bit(size));
```
(Note: this requires the `<bit>` header to be included as well)

Reply:
Nice catch!